From 50028085f638264e7a080d9be751a1ceb9ceb013 Mon Sep 17 00:00:00 2001
From: Nikita Popov <nikita.ppv@gmail.com>
Date: Mon, 15 Feb 2021 15:33:42 +0100
Subject: [PATCH] Fix #80747: Providing RSA key size < 512 generates key that
 crash PHP

When RSA_generate_key_ex() fails (e.g. fewer than 512 bits) the empty RSA was still
assigned to the key, so openssl_pkey_new() returned an unusable key instead of false.

Origin: backport, https://github.com/php/php-src/commit/64b10854643531bcf62a884aaea0bafc57e4f914
Bug: https://bugs.php.net/80747
---
 ext/openssl/openssl.c           |  2 ++
 ext/openssl/tests/bug80747.phpt | 18 ++++++++++++++++++
 2 files changed, 20 insertions(+)
 create mode 100644 ext/openssl/tests/bug80747.phpt

diff --git a/ext/openssl/openssl.c b/ext/openssl/openssl.c
index aec8bb75..e75da828 100644
--- a/ext/openssl/openssl.c
+++ b/ext/openssl/openssl.c
@@ -3946,6 +3946,8 @@ static EVP_PKEY * php_openssl_generate_private_key(struct php_x509_request * req
 						PHP_OPENSSL_RAND_ADD_TIME();
 						if (rsaparam == NULL || !RSA_generate_key_ex(rsaparam, req->priv_key_bits, bne, NULL)) {
 							php_openssl_store_errors();
+							RSA_free(rsaparam);
+							rsaparam = NULL;
 						}
 						BN_free(bne);
 					}
diff --git a/ext/openssl/tests/bug80747.phpt b/ext/openssl/tests/bug80747.phpt
new file mode 100644
index 00000000..049405ec
--- /dev/null
+++ b/ext/openssl/tests/bug80747.phpt
@@ -0,0 +1,18 @@
+--TEST--
+Bug #80747: Providing RSA key size < 512 generates key that crash PHP
+--SKIPIF--
+<?php
+if (!extension_loaded("openssl")) die("skip");
+?>
+--FILE--
+<?php
+
+$conf = array(
+    'config' => __DIR__ . DIRECTORY_SEPARATOR . 'openssl.cnf',
+    'private_key_bits' => 511,
+);
+var_dump(openssl_pkey_new($conf));
+
+?>
+--EXPECT--
+bool(false)
